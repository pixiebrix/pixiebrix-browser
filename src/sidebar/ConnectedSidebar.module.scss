/*
 * Copyright (C) 2022 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

@import "@/themes/colors";

// Due to design system inconsistencies,
// we have to override our primary purple color
// for the PixieBrix theme
.themeColorOverride {
  color: $N800;

  &:hover {
    color: $N300;
  }
}

.themeColor {
  color: var(--theme-primary);

  &:hover {
    color: var(--theme-primary);
    background-color: rgba(0, 0, 0, 0.07);
  }
}

.button {
  display: inline-flex;
  align-items: center;
  text-decoration: none;
}

.logo {
  // see _themes.scss
  height: var(--logo-height);
  padding: 0 1.5rem;
}
