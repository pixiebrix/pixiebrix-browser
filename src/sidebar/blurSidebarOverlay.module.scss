/*!
 * Copyright (C) 2024 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

.blurOverlay {
  display: flex;
  justify-content: center;
  align-items: flex-start;
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background-color: rgba(255, 255, 255, 0.5);
  backdrop-filter: blur(2.5px);
  pointer-events: all;
  z-index: 1; // ensure overlay is on top of the sidebar panel content
  > div {
    margin-top: 15vh;
    width: 300px;
  }
}

.modalDialog {
  text-align: center;
  margin: 23px;
  padding: 16px;
  border-radius: 12px;
  background: white;
  border: 1px solid #cfcbd6;
  > div {
    border: 0;
  }
}

.modalHeader {
  display: block;
  padding-bottom: 0;
  background: white;
  border: 0;
}

.modalBody {
  background: white;
  border: 0;
}
