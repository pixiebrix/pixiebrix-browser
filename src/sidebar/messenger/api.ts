/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

/* Do not use `registerMethod` in this file */
import { getMethod, getNotifier } from "webext-messenger";

const target = { tabId: "this", page: "/sidebar.html" } as const;
const sidebarInThisTab = {
  renderPanels: getMethod("SIDEBAR_RENDER_PANELS", target),
  activatePanel: getMethod("SIDEBAR_ACTIVATE_PANEL", target),
  showForm: getMethod("SIDEBAR_SHOW_FORM", target),
  hideForm: getMethod("SIDEBAR_HIDE_FORM", target),
  pingSidebar: getMethod("SIDEBAR_PING", target),
  showTemporaryPanel: getMethod("SIDEBAR_SHOW_TEMPORARY_PANEL", target),
  updateTemporaryPanel: getNotifier("SIDEBAR_UPDATE_TEMPORARY_PANEL", target),
  hideTemporaryPanel: getMethod("SIDEBAR_HIDE_TEMPORARY_PANEL", target),
  showModActivationPanel: getMethod("SIDEBAR_SHOW_ACTIVATE_RECIPE", target),
  hideModActivationPanel: getMethod("SIDEBAR_HIDE_ACTIVATE_RECIPE", target),
};

export default sidebarInThisTab;
