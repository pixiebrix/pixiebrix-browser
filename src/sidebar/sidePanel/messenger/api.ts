/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

/**
 * @file This file defines the public API for the sidePanel, with some
 * exceptions that use `expectContext`. It uses the `messenger/api.ts` name
 * to match that expectation and avoid lint issues.
 */

import { getExtensionConsoleUrl } from "@/utils/extensionUtils";
import {
  DISPLAY_REASON_EXTENSION_CONSOLE,
  DISPLAY_REASON_RESTRICTED_URL,
} from "@/tinyPages/restrictedUrlPopupConstants";
import { isScriptableUrl } from "webext-content-scripts";
import { isObject } from "@/utils/objectUtils";
import { expectContext } from "@/utils/expectContext";
import { type Target } from "webext-messenger";

export function getAssociatedTabId(): number {
  expectContext("sidebar");
  const tabId = new URLSearchParams(window.location.search).get("tabId");
  return Number(tabId);
}

export function getAssociatedTarget(): Target {
  return { tabId: getAssociatedTabId(), frameId: 0 };
}

const PING_MESSAGE = "PING_SIDE_PANEL";
// Do not use the messenger because it doesn't support retry-less messaging
// TODO: Drop after https://github.com/pixiebrix/webext-messenger/issues/59
export function respondToPings() {
  expectContext("sidebar");
  chrome.runtime.onMessage.addListener((message, sender, sendResponse) => {
    if (
      isObject(message) &&
      message.type === PING_MESSAGE &&
      sender.tab?.id === getAssociatedTabId()
    ) {
      sendResponse(true);
    }
  });
}

export async function isSidePanelOpen(): Promise<boolean> {
  // Sync check where possible
  if (isSidePanelOpenSync() === false) {
    return false;
  }

  const response = await chrome.runtime.sendMessage<
    { type: string },
    boolean | undefined
  >({ type: PING_MESSAGE });
  return Boolean(response); // TODO: Drop Boolean() after strictNullChecks migration
}

export function getPopoverUrl(tabUrl: string | undefined): string | null {
  const popoverUrl = browser.runtime.getURL("restrictedUrlPopup.html");

  if (tabUrl?.startsWith(getExtensionConsoleUrl())) {
    return `${popoverUrl}?reason=${DISPLAY_REASON_EXTENSION_CONSOLE}`;
  }

  if (!isScriptableUrl(tabUrl)) {
    return `${popoverUrl}?reason=${DISPLAY_REASON_RESTRICTED_URL}`;
  }

  // The popup is disabled, and the extension will receive browserAction.onClicked events.
  // https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/browserAction/setPopup#popup
  return null;
}

export function getSidebarPath(tabId: number, url: string): string {
  return getPopoverUrl(url) ?? "sidebar.html?tabId=" + tabId;
}

export async function openSidePanel(tabId: number, url: string) {
  // Simultaneously define, enable, and open the side panel
  void chrome.sidePanel.setOptions({
    tabId,
    path: getSidebarPath(tabId, url),
    enabled: true,
  });
  await chrome.sidePanel.open({ tabId });
}

export async function hideSidePanel(tabId: number) {
  void chrome.sidePanel.setOptions({
    tabId,
    enabled: false,
  });
}

// Approximate sidebar width in pixels. Used to determine whether it's open
const MINIMUM_SIDEBAR_WIDTH = 300;

/**
 * Determines whether the sidebar is open.
 * @returns false when it's definitely closed
 * @returns 'unknown' when it cannot be determined
 */
// The type cannot be `undefined` due to strictNullChecks
function isSidePanelOpenSync(): false | "unknown" {
  if (!globalThis.window) {
    return "unknown";
  }

  return window.outerWidth - window.innerWidth > MINIMUM_SIDEBAR_WIDTH
    ? "unknown"
    : false;
}

// TODO: It doesn't work when the dev tools are open on the side
// Official event requested in https://github.com/w3c/webextensions/issues/517
export function onSidePanelClosure(controller: AbortController): void {
  expectContext("contentScript");
  window.addEventListener(
    "resize",
    () => {
      if (isSidePanelOpenSync() === false) {
        controller.abort();
      }
    },
    { signal: controller.signal },
  );
}
