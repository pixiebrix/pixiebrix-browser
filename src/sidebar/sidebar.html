<!--
  ~ Copyright (C) 2022 PixieBrix, Inc.
  ~
  ~ This program is free software: you can redistribute it and/or modify
  ~ it under the terms of the GNU Affero General Public License as published by
  ~ the Free Software Foundation, either version 3 of the License, or
  ~ (at your option) any later version.
  ~
  ~ This program is distributed in the hope that it will be useful,
  ~ but WITHOUT ANY WARRANTY; without even the implied warranty of
  ~ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  ~ GNU Affero General Public License for more details.
  ~
  ~ You should have received a copy of the GNU Affero General Public License
  ~ along with this program.  If not, see <http://www.gnu.org/licenses/>.
  -->

<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <!-- Open all links in new tab, never in the sidebar #851 -->
    <base target="_blank" />
    <title>PixieBrix: Page Panel</title>
    <link rel="stylesheet" href="loadingScreen.css" />
    <link rel="stylesheet" href="sidebar.css" />
  </head>
  <body>
    <div id="container" class="full-height vh-100">
      <div class="global-loading-message">Loading Sidebar ðŸš€</div>
    </div>
    <script src="loadingScreen.js"></script>
    <script src="sidebar.js"></script>
    <!-- Must be at the bottom https://github.com/pixiebrix/pixiebrix-extension/issues/4100 -->
    <link
      href="https://fonts.googleapis.com/css2?family=Ubuntu:wght@300;400;500;700&display=swap"
      rel="stylesheet"
    />
  </body>
</html>
