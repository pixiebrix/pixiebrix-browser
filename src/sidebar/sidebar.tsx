/*
 * Copyright (C) 2022 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

// Keep in order so precedence is preserved
import "@/vendors/theme/app/app.scss";
import "@/vendors/overrides.scss";
import "@/utils/layout.scss";
import "./sidebar.scss";

import "@/extensionContext";

import registerMessenger from "@/sidebar/messenger/registration";
import App from "@/sidebar/SidebarApp";
import ReactDOM from "react-dom";
import React from "react";
import registerBuiltinBlocks from "@/blocks/registerBuiltinBlocks";
import registerContribBlocks from "@/contrib/registerContribBlocks";
import { initToaster } from "@/utils/notify";

registerMessenger();
registerContribBlocks();
registerBuiltinBlocks();
initToaster();

ReactDOM.render(<App />, document.querySelector("#container"));
