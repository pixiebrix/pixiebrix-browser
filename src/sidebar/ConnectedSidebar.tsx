/*
 * Copyright (C) 2022 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import styles from "./ConnectedSidebar.module.scss";

import React, { Dispatch, useEffect, useMemo } from "react";
import { Button } from "react-bootstrap";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faAngleDoubleRight, faCog } from "@fortawesome/free-solid-svg-icons";
import {
  addListener,
  removeListener,
  SidebarListener,
} from "@/sidebar/protocol";
import DefaultPanel from "@/sidebar/DefaultPanel";
import { useDispatch, useSelector } from "react-redux";
import { ActivatePanelOptions, FormEntry, PanelEntry } from "@/sidebar/types";
import Tabs from "@/sidebar/Tabs";
import sidebarSlice, { SidebarState } from "./sidebarSlice";
import { AnyAction } from "redux";
import { hideSidebar } from "@/contentScript/messenger/api";
import { whoAmI } from "@/background/messenger/api";
import useTheme from "@/hooks/useTheme";
import RequireAuth from "@/auth/RequireAuth";
import LoginPanel from "@/sidebar/LoginPanel";

/**
 * Listeners to update the Sidebar's Redux state upon receiving messages from the contentScript.
 */
function getConnectedListener(dispatch: Dispatch<AnyAction>): SidebarListener {
  return {
    onRenderPanels(panels: PanelEntry[]) {
      dispatch(sidebarSlice.actions.setPanels({ panels }));
    },
    onShowForm(form: FormEntry) {
      dispatch(sidebarSlice.actions.addForm({ form }));
    },
    onHideForm({ nonce }: Partial<FormEntry>) {
      dispatch(sidebarSlice.actions.removeForm(nonce));
    },
    onActivatePanel(options: ActivatePanelOptions) {
      dispatch(sidebarSlice.actions.activatePanel(options));
    },
  };
}

const selectState = ({ sidebar }: { sidebar: SidebarState }) => sidebar;

const ConnectedSidebar: React.VFC = () => {
  const { logo } = useTheme();
  const dispatch = useDispatch();
  const sidebarState = useSelector(selectState);

  const listener: SidebarListener = useMemo(
    () => getConnectedListener(dispatch),
    [dispatch]
  );

  // `effect` will run once on component mount since listener and formsRef don't change on renders
  useEffect(() => {
    addListener(listener);
    return () => {
      // NOTE: we don't need to cancel any outstanding forms on unmount because the FormTransformer is set up to watch
      // for PANEL_HIDING_EVENT. (and the only time this SidebarApp would unmount is if the sidebar was closing)
      removeListener(listener);
    };
  }, [listener]);

  return (
    <div className="full-height">
      <div className="d-flex p-2 justify-content-between align-content-center">
        <Button
          className={styles.button}
          onClick={async () => {
            const sidebar = await whoAmI();
            await hideSidebar({ tabId: sidebar.tab.id });
          }}
          size="sm"
          variant="link"
        >
          <FontAwesomeIcon icon={faAngleDoubleRight} className="fa-lg" />
        </Button>
        <div className="align-self-center">
          <img
            src={logo.regular}
            alt="PixieBrix logo"
            className={styles.logo}
          />
        </div>
        <Button
          href="/options.html"
          target="_blank"
          size="sm"
          variant="link"
          className={styles.button}
        >
          <span>
            Options <FontAwesomeIcon icon={faCog} />
          </span>
        </Button>
      </div>

      <div className="full-height">
        <RequireAuth LoginPage={LoginPanel}>
          {sidebarState.panels?.length || sidebarState.forms?.length ? (
            <Tabs
              {...sidebarState}
              onSelectTab={(eventKey: string) => {
                dispatch(sidebarSlice.actions.selectTab(eventKey));
              }}
            />
          ) : (
            <DefaultPanel />
          )}
        </RequireAuth>
      </div>
    </div>
  );
};

export default ConnectedSidebar;
