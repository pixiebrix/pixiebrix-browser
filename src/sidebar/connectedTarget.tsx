/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { isMV3 } from "@/mv3/api";
import { expectContext } from "@/utils/expectContext";
import { assertNotNullish } from "@/utils/nullishUtils";
import { once } from "lodash";
import { type TopLevelFrame, getTopLevelFrame } from "webext-messenger";
import { getTabUrl } from "webext-tools";

function getConnectedTabIdMv3(): number {
  expectContext("sidebar");
  const tabId = new URLSearchParams(window.location.search).get("tabId");
  assertNotNullish(
    tabId,
    `No tabId argument was found on this page: ${window.location.href}`,
  );
  return Number(tabId);
}

async function getConnectedTabIdMv2() {
  const { tabId } = await getTopLevelFrame();
  return tabId;
}

export const getConnectedTabId = once(
  isMV3() ? getConnectedTabIdMv3 : getConnectedTabIdMv2,
);
export const getConnectedTarget = isMV3()
  ? (): TopLevelFrame => ({ tabId: getConnectedTabIdMv3(), frameId: 0 })
  : getTopLevelFrame;

export async function getAssociatedTargetUrl(): Promise<string | undefined> {
  return getTabUrl(await getConnectedTarget());
}
