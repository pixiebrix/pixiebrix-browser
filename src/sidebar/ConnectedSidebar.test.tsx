/*
 * Copyright (C) 2022 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React from "react";
import ConnectedSidebar from "@/sidebar/ConnectedSidebar";
import { render } from "@/sidebar/testHelpers";
import { authActions } from "@/auth/authSlice";
import { authStateFactory } from "@/testUtils/factories";
import { waitForEffect } from "@/testUtils/testHelpers";
import { useGetMeQuery } from "@/services/api";
import { anonAuth } from "@/auth/authConstants";
import { MemoryRouter } from "react-router";

jest.mock("@/store/optionsStore", () => ({
  persistor: {
    flush: jest.fn(),
  },
}));
jest.mock("@/services/api", () => ({
  useGetMeQuery: jest.fn(),
  appApi: {
    reducerPath: "appApi",
  },
}));
jest.mock("@/auth/token", () => {
  const originalModule = jest.requireActual("@/auth/token");
  return {
    ...originalModule,
    isLinked: jest.fn().mockResolvedValue(true),
  };
});
browser.runtime.getURL = (path: string) =>
  `chrome-extension://example.url/${path}`;

beforeAll(() => {
  jest.useFakeTimers();
});

afterAll(() => {
  jest.runAllTimers();
  jest.useRealTimers();
});

describe("SidebarApp", () => {
  test("renders not connected", async () => {
    (useGetMeQuery as jest.Mock).mockReturnValue({
      isLoading: false,
      data: anonAuth,
    });

    const rendered = render(
      <MemoryRouter>
        <ConnectedSidebar />
      </MemoryRouter>
    );
    await waitForEffect();

    jest.runAllTimers();
    expect(rendered.asFragment()).toMatchSnapshot();
  });

  test("renders not connected partner view", async () => {
    (useGetMeQuery as jest.Mock).mockReturnValue({
      isLoading: false,
      data: {
        partner: {},
        ...anonAuth,
      },
    });

    const rendered = render(
      <MemoryRouter>
        <ConnectedSidebar />
      </MemoryRouter>
    );
    await waitForEffect();

    jest.runAllTimers();
    expect(rendered.asFragment()).toMatchSnapshot();
  });

  test("renders", async () => {
    (useGetMeQuery as jest.Mock).mockReturnValue({
      isLoading: false,
      data: authStateFactory(),
    });

    const rendered = render(
      <MemoryRouter>
        <ConnectedSidebar />
      </MemoryRouter>,
      {
        setupRedux(dispatch) {
          dispatch(authActions.setAuth(authStateFactory()));
        },
      }
    );

    await waitForEffect();

    jest.runAllTimers();
    expect(rendered.asFragment()).toMatchSnapshot();
  });
});
