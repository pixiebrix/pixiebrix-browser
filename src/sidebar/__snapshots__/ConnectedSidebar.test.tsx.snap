// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SidebarApp renders 1`] = `
<DocumentFragment>
  <div
    class="full-height"
  >
    <div>
      <div
        class="container"
      >
        <div
          class="mt-4 row"
        >
          <div
            class="col"
          >
            <h4
              class="display-6"
            >
              Activate an Official Blueprint
            </h4>
            <p>
              <span
                class="text-primary"
              >
                The easiest way to start using PixieBrix!
              </span>
               Activate a pre-made blueprint from the Public Marketplace.
            </p>
            <a
              class="btn btn-info"
              href="https://www.pixiebrix.com/marketplace/"
              rel="noopener noreferrer"
              target="_blank"
            >
              Open Marketplace 
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-external-link-alt fa-w-16 "
                data-icon="external-link-alt"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M432,320H400a16,16,0,0,0-16,16V448H64V128H208a16,16,0,0,0,16-16V80a16,16,0,0,0-16-16H48A48,48,0,0,0,0,112V464a48,48,0,0,0,48,48H400a48,48,0,0,0,48-48V336A16,16,0,0,0,432,320ZM488,0h-128c-21.37,0-32.05,25.91-17,41l35.73,35.73L135,320.37a24,24,0,0,0,0,34L157.67,377a24,24,0,0,0,34,0L435.28,133.32,471,169c15,15,41,4.5,41-17V24A24,24,0,0,0,488,0Z"
                  fill="currentColor"
                />
              </svg>
            </a>
          </div>
        </div>
        <div
          class="mt-4 row"
        >
          <div
            class="col"
          >
            <h4
              class="display-6"
            >
              Create your Own
            </h4>
            <p>
              Follow the Quick Start Guide to start creating your own bricks in minutes.
            </p>
            <a
              class="btn btn-info"
              href="https://docs.pixiebrix.com/quick-start-guide"
              rel="noopener noreferrer"
              target="_blank"
            >
              Open Quick Start Guide 
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-external-link-alt fa-w-16 "
                data-icon="external-link-alt"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M432,320H400a16,16,0,0,0-16,16V448H64V128H208a16,16,0,0,0,16-16V80a16,16,0,0,0-16-16H48A48,48,0,0,0,0,112V464a48,48,0,0,0,48,48H400a48,48,0,0,0,48-48V336A16,16,0,0,0,432,320ZM488,0h-128c-21.37,0-32.05,25.91-17,41l35.73,35.73L135,320.37a24,24,0,0,0,0,34L157.67,377a24,24,0,0,0,34,0L435.28,133.32,471,169c15,15,41,4.5,41-17V24A24,24,0,0,0,488,0Z"
                  fill="currentColor"
                />
              </svg>
            </a>
          </div>
        </div>
        <div
          class="row"
        >
          <div
            class="text-center col"
          >
            <img
              alt="Marketplace"
              src="test-file-stub"
              width="300"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`SidebarApp renders not connected 1`] = `
<DocumentFragment>
  <div
    class="full-height"
  >
    <div
      class="container"
    >
      <div
        class="mt-4 row"
      >
        <div
          class="text-center col"
        >
          <h4
            class="display-6"
          >
            Connect PixieBrix Account
          </h4>
          <p>
            Register/log-in to PixieBrix to access your personal and team bricks
          </p>
          <a
            class="mt-4 btn btn-primary"
            href="https://app.pixiebrix.com"
            target="_blank"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-sign-in-alt fa-w-16 "
              data-icon="sign-in-alt"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M416 448h-84c-6.6 0-12-5.4-12-12v-40c0-6.6 5.4-12 12-12h84c17.7 0 32-14.3 32-32V160c0-17.7-14.3-32-32-32h-84c-6.6 0-12-5.4-12-12V76c0-6.6 5.4-12 12-12h84c53 0 96 43 96 96v192c0 53-43 96-96 96zm-47-201L201 79c-15-15-41-4.5-41 17v96H24c-13.3 0-24 10.7-24 24v96c0 13.3 10.7 24 24 24h136v96c0 21.5 26 32 41 17l168-168c9.3-9.4 9.3-24.6 0-34z"
                fill="currentColor"
              />
            </svg>
             Connect Account
          </a>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="text-center col"
        >
          <img
            alt="Marketplace"
            src="test-file-stub"
            width="300"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
